<div class="Body u-font-serif" id="body"><div><section id="sec0005"><h2 class="u-h3 u-margin-l-top u-margin-xs-bottom">1. Introduction</h2><p id="par0005">Complex reaction networks are at the core of numerous chemical and biochemical systems (<a name="bbib0065" href="#bib0065" class="workspace-trigger">Broadbelt &amp; Pfaendtner, 2005</a>); examples include pyrolysis (<a name="bbib0145" href="#bib0145" class="workspace-trigger">Froment et al., 1976</a>, <a name="bbib0240" href="#bib0240" class="workspace-trigger">Mohan et al., 2006</a>) and acid-catalyzed (<a name="bbib0100" href="#bib0100" class="workspace-trigger">Corma et al., 2008</a>, <a name="bbib0385" href="#bib0385" class="workspace-trigger">Yaluris et al., 1997</a>) conversion of hydrocarbons or oxygenates, and biochemical networks such as glycolysis in <em>Saccharomyces cerevisiae</em> (<a name="bbib0320" href="#bib0320" class="workspace-trigger">Rizzi, Baltes, Theobald, &amp; Reuss, 1997</a>) and genome-scale models of <em>Escherichia coli</em> (<a name="bbib0310" href="#bib0310" class="workspace-trigger">Reed, Vo, Schilling, &amp; Palsson, 2003</a>). Such networks are also extensively prevalent in industrial reaction systems, particularly in petroleum refining, as the crude oil feed is a mixture of many compounds and a multitude of chemical transformations can potentially take place (<a name="bbib0180" href="#bib0180" class="workspace-trigger">Ho, 2008</a>, <a name="bbib0285" href="#bib0285" class="workspace-trigger">Quann and Jaffe, 1996</a>). Despite the large size, the reaction networks can be constructed from a relatively smaller set of chemical transformation rules. Automated network generators have been developed to construct such networks from initial reactants and pre-specified reaction rules.</p><p id="par0010">Kinetic modeling is a common application of automated network generation, wherein the appropriate differential algebraic system of equations that captures the dynamics of the system is formulated. The model is then solved with kinetic parameters estimated, predicted, or specified, to obtain product yield information (<a name="bbib0105" href="#bib0105" class="workspace-trigger">Cortright &amp; Dumesic, 2001</a>). Network generation in conjunction with kinetic modeling has been extensively applied for hydrocarbon (<a name="bbib0045" href="#bib0045" class="workspace-trigger">Broadbelt et al., 1994a</a>, <a name="bbib0360" href="#bib0360" class="workspace-trigger">Van Geem et al., 2006</a>) and biochemical systems (<a name="bbib0395" href="#bib0395" class="workspace-trigger">Faeder et al., 2003</a>). Complex reaction networks, however, have also been analyzed for: (a) deriving topological properties such as average path length of the network (<a name="bbib0010" href="#bib0010" class="workspace-trigger">Arita, 2003</a>, <a name="bbib0015" href="#bib0015" class="workspace-trigger">Arita, 2004</a>), (b) identifying synthetic/degradation pathways (<a name="bbib0400" href="#bib0400" class="workspace-trigger">Gonzalez-Lergier, Broadbelt, &amp; Hatzimanikatis, 2005</a>), and (c) deriving and testing plausible mechanisms and overall rate expressions (<a name="bbib0125" href="#bib0125" class="workspace-trigger">Fan et al., 2002</a>, <a name="bbib0140" href="#bib0140" class="workspace-trigger">Fishtik et al., 2004</a>, <a name="bbib0220" href="#bib0220" class="workspace-trigger">Lin et al., 2009</a>). The use of additional thermodynamic data in conjunction with the reaction network has further enabled quantitative analysis of networks in terms of: (a) generating thermodynamically meaningful flux distributions in biochemical systems (<a name="bbib0160" href="#bib0160" class="workspace-trigger">Henry, Broadbelt, &amp; Hatzimanikatis, 2007</a>), (b) extracting functional information such as regulatory sites in biological systems (<a name="bbib0200" href="#bib0200" class="workspace-trigger">Kummel, Panke, &amp; Heinemann, 2006</a>), and (c) identifying thermodynamically feasible synthesis routes (<a name="bbib0210" href="#bib0210" class="workspace-trigger">Li et al., 2004</a>) to form chemicals, or biological degradation pathways (<a name="bbib0135" href="#bib0135" class="workspace-trigger">Finley, Broadbelt, &amp; Hatzimanikatis, 2009</a>) to decompose molecules.</p><p id="par0015">All these examples require a constructed network as the first step and graph theoretical analysis of the network in subsequent steps; hence, an automated network generator can be used to construct a complex network prior to network analysis. In this context, we have developed a new computational tool, Rule Input Network Generator, or RING (<a name="bbib0290" href="#bib0290" class="workspace-trigger">Rangarajan, Bhan, &amp; Daoutidis, 2010</a>), for network generation and analysis of complex thermochemical reaction systems. RING takes initial reactants, chemistry rules and network analysis instructions as inputs written in a domain specific English-like reaction language. RING then generates an exhaustive network of reactions and analyzes the network in a post-processing step to obtain: (a) lumps of isomeric chemical species in the network, (b) reaction pathways from reactants to products, and (c) mechanisms that describe the overall chemical transformation of the initial reactants to specified products. RING is available open source (<a name="bbib0315" href="#bib0315" class="workspace-trigger">RING, 2011</a>) under GNU Lesser GPL.</p><p id="par0020">In our previous work (<a name="bbib0290" href="#bib0290" class="workspace-trigger">Rangarajan et al., 2010</a>), the reaction generation feature in RING was introduced briefly and its application to biomass conversion was illustrated through four relevant reaction systems as examples. We have subsequently developed a domain specific English-like extensible reaction language as a user interface and a post-processing module, as discussed above. In this article, the first of a two-part series, we discuss all features of RING with illustrative examples. Applications of RING in complex network analysis of catalytic systems are discussed in the second part.</p></section><section id="sec0010"><h2 class="u-h3 u-margin-l-top u-margin-xs-bottom">2. Network generation and analysis: a review</h2><section id="sec0015"><h3 class="u-h4 u-margin-m-top u-margin-xs-bottom">2.1. Network generators</h3><div><p id="par0025">Rule-based automated reaction network generators are computational tools that take in a set of molecules as reactants, and iteratively apply a set of reaction rules, to construct a comprehensive list of possible reactions. Network generators have been developed and applied in different fields such as pyrolysis and oxidation, catalysis, and biological systems. <a name="btbl0005" href="#tbl0005" class="workspace-trigger">Table 1</a> lists and describes several of them. All automated network generators have five common and essential features (<a name="bbib0350" href="#bib0350" class="workspace-trigger">Tomlin et al., 1997, chap. 4</a>). First, an unambiguous representation is required for molecules and reactions. This is usually represented as character strings for input and output (<a name="bbib0045" href="#bib0045" class="workspace-trigger">Broadbelt et al., 1994a</a>, <a name="bbib0265" href="#bib0265" class="workspace-trigger">Prickett and Mavrovouniotis, 1997a</a>, <a name="bbib0305" href="#bib0305" class="workspace-trigger">Ratkiewicz and Truong, 2003</a>). Second, an internal representation of molecules is required, such as molecular trees, adjacency matrices, or chemical graphs, thereby enabling quick structure manipulation. Adjacency matrices are the most common representation format owing to their simplicity. An adjacency matrix ‘<em>M</em>’ of a molecule is a square matrix containing connectivity and bond order information between every two atoms. Thus, <em>M</em>(<em>i</em>,<em>j</em>)&nbsp;=&nbsp;0 implies that the <em>i</em>th and <em>j</em>th atoms are not connected while a positive nonzero value would indicate the strength of the bond (1 is a single bond, 2 is a double bond, etc.). The diagonal values indicate the number of unpaired electrons in the atoms. Third, an internal representation of reaction rules that can be applied iteratively on the molecules is required. A common representation scheme is to employ a matrix for reaction rules. Proposed initially by Dugundji and Ugi (<a name="bbib0120" href="#bib0120" class="workspace-trigger">Dugundji &amp; Ugi, 1973</a>, also <a name="bbib0355" href="#bib0355" class="workspace-trigger">Ugi et al., 1993</a>) and later used in other tools such as NETGEN (<a name="bbib0045" href="#bib0045" class="workspace-trigger">Broadbelt et al., 1994a</a>), the reaction matrix describes the transformations for an adjacency matrix representation of molecules. The reaction matrix is structured similarly, with off-diagonal values representing bond changes and diagonal entries representing changes in unpaired electron count. Further, positive/negative values indicate increase/decrease in bond order, bond formation/cleavage, or increase/decrease in unpaired electrons of the atom. <a name="bbib0020" href="#bib0020" class="workspace-trigger">Baltanas and Froment (1985)</a> used a Boolean matrix to represent molecules for generation of networks for modeling paraffin cracking and isomerization on bifunctional catalysts. The Boolean matrix is similar to the adjacency matrix; however, bonds of a higher order (e.g. double bonds) and information on charges (such as +1 for carbenium ions) are stored separately. This method, therefore, is similar to that of <a name="bbib0120" href="#bib0120" class="workspace-trigger">Dugundji and Ugi (1973)</a>. Transformations in RDL (<a name="bbib0265" href="#bib0265" class="workspace-trigger">Prickett &amp; Mavrovouniotis, 1997a</a>) and RDL++ (<a name="bbib0185" href="#bib0185" class="workspace-trigger">Hsu et al., 2008</a>), on the other hand, are input by the user as English-language-like statements describing changes in the charge/bonding of atoms participating in the reaction rule which get directly applied on the internal graph description of molecules. Fourth, all network generators have a generation scheme that iteratively applies the reaction rules to all input and generated molecules so that the resultant network is exhaustive. The scheme should ensure that all possible reactions of a given set of reactants are generated corresponding to that reaction rule. <a name="bbib0130" href="#bib0130" class="workspace-trigger">Faulon and Sault (2001)</a> describe such a generation scheme as deterministic network generation.</p><div class="tables frame-topbot rowsep-0 colsep-0" id="tbl0005"><span class="captions"><span><p id="spar0055"><span class="label">Table 1</span>. A list of reaction network generators, and a description of their essential features, and their areas of application.</p></span></span><div class="groups"><table><thead class="valign-top"><tr><th scope="col" class="rowsep-1 align-left">Name</th><th scope="col" class="rowsep-1 align-left">Description</th><th scope="col" class="rowsep-1 align-left">Remarks</th><th scope="col" class="rowsep-1 align-left">References</th></tr></thead><tbody><tr><td class="align-left">NETGEN</td><td class="align-left">Network generator and model builder based on ‘BE’ &amp; ‘R’ matrices. Uses adjacency matrix representation of molecules and reaction rules.</td><td class="align-left">(i) Rate based and rank based pruning. (ii) Application in gas phase pyrolysis, nanoparticle synthesis, and biochemical reactions. (iii) Linked to <a name="bbib0245" href="#bib0245" class="workspace-trigger">MOPAC2009 (2008)</a> and NIST S&amp;P database for thermochemistry.</td><td class="align-left"><a name="bbib0210" href="#bib0210" class="workspace-trigger">Li et al. (2004)</a>, <a name="bbib0045" href="#bib0045" class="workspace-trigger">Broadbelt et al. (1994a)</a>, <a name="bbib0050" href="#bib0050" class="workspace-trigger">Broadbelt et al., 1994b</a>, <a name="bbib0055" href="#bib0055" class="workspace-trigger">Broadbelt et al., 1995</a>, <a name="bbib0060" href="#bib0060" class="workspace-trigger">Broadbelt et al., 1996</a>, <a name="bbib0345" href="#bib0345" class="workspace-trigger">Susnow, Dean, Green, Peczak, and Broadbelt (1997)</a>, <a name="bbib0370" href="#bib0370" class="workspace-trigger">Wong, Li, Swihart, and Broadbelt (2004)</a>, <a name="bbib0260" href="#bib0260" class="workspace-trigger">Pfaendtner and Broadbelt (2008)</a></td></tr><tr><td class="align-left">EXGAS</td><td class="align-left">Kinetic model builder using a tree data structure for internal molecule representation.</td><td class="align-left">(i) Applied in gas phase combustion and oxidation. (ii) A library of free radical chemistry rules used in reaction network generation. (iii) Tree representation system based on <a name="bbib0085" href="#bib0085" class="workspace-trigger">Chinnick, Baulch, and Asyscough (1988)</a>.</td><td class="align-left"><a name="bbib0170" href="#bib0170" class="workspace-trigger">Heyberger et al. (2001)</a>, <a name="bbib0375" href="#bib0375" class="workspace-trigger">Warth et al. (2000)</a></td></tr><tr><td class="align-left">COMGEN</td><td class="align-left">Network generator based on chemical graph theory. String representation of molecules, reactant pattern based on <a name="bbib0040" href="#bib0040" class="workspace-trigger">Blurock (1995)</a>, and topological indices for molecule identification.</td><td class="align-left">(i) Hydrocarbon gas phase chemistry. (ii) Thermochemistry was calculated from a database.</td><td class="align-left"><a name="bbib0305" href="#bib0305" class="workspace-trigger">Ratkiewicz and Truong (2003)</a></td></tr><tr><td class="align-left">RMG</td><td class="align-left">Kinetic models of free radical chemistry of hydrocarbons. Kinetics estimated from semi-empirical relations obtained from theoretical calculations.</td><td class="align-left">(i) Applications in hydrocarbon pyrolysis. (ii) Accurate calculations of kinetics and thermodynamics, and formulation of kinetic models to predict product yields and conversion.</td><td class="align-left"><a name="bbib0340" href="#bib0340" class="workspace-trigger">Song (2004)</a>, <a name="bbib0360" href="#bib0360" class="workspace-trigger">Van Geem et al. (2006)</a>, <a name="bbib0150" href="#bib0150" class="workspace-trigger">Green et al. (2004)</a></td></tr><tr><td class="align-left">RDL</td><td class="align-left">English-like language based description of reaction rules. Object-oriented framework using elements of graph theory.</td><td class="align-left">(i) Reaction network generated from scratch depending upon reaction rules input, thus offering flexibility in describing the system. (ii) Constraints on rules to prevent combinatorial explosion.</td><td class="align-left"><a name="bbib0265" href="#bib0265" class="workspace-trigger">Prickett and Mavrovouniotis, 1997a</a>, <a name="bbib0270" href="#bib0270" class="workspace-trigger">Prickett and Mavrovouniotis, 1997b</a>, <a name="bbib0275" href="#bib0275" class="workspace-trigger">Prickett and Mavrovouniotis, 1997c</a></td></tr><tr><td class="align-left">RDL++</td><td class="align-left">Extends RDL with additional features to enable description of solid-acid catalyzed reactions of hydrocarbons.</td><td class="align-left">(i) Applied in microkinetic modeling of heterogeneous catalytic systems, and data analysis and knowledge extraction in high-throughput experimentation.</td><td class="align-left"><a name="bbib0185" href="#bib0185" class="workspace-trigger">Hsu et al. (2008)</a>, <a name="bbib0195" href="#bib0195" class="workspace-trigger">Katare, Caruthers, Delgass, and Venkatasubramanian (2004)</a>, <a name="bbib0025" href="#bib0025" class="workspace-trigger">Bhan et al. (2005)</a>, <a name="bbib0075" href="#bib0075" class="workspace-trigger">Caruthers et al. (2003)</a></td></tr><tr><td class="align-left">KING</td><td class="align-left">An automated mechanism generator. Uses ‘BE’ and ‘R’ matrix for molecule and reaction representation.</td><td class="align-left">(i) Applied in combustion chemistry. (ii) Reactions are determined combinatorially, as a linear combination of elementary steps.</td><td class="align-left"><a name="bbib0115" href="#bib0115" class="workspace-trigger">Di Maio and Lignola (1992)</a></td></tr><tr><td class="align-left">Bio-NETGEN</td><td class="align-left">Rule-based generation of biological reaction network. Graph based representation of molecules with each node being a building block of the macromolecule of the biological system.</td><td class="align-left">(i) Application in reaction network generation in biological systems and subsequent dynamic modeling.</td><td class="align-left"><a name="bbib0030" href="#bib0030" class="workspace-trigger">Blinov, Yang, Faeder, and Hlavacek (2005)</a>, <a name="bbib0035" href="#bib0035" class="workspace-trigger">Blinov, Faeder, Yang, Goldstein, and Hlavacek (2005)</a>, <a name="bbib0390" href="#bib0390" class="workspace-trigger">Faeder, Blinov, Goldstein, and Hlavacek (2005)</a></td></tr><tr><td class="align-left">BNICE</td><td class="align-left">Computational framework for generating and analyzing biological reaction pathways.</td><td class="align-left">(i) Reaction rules are obtained from the enzyme function information in KEGG database. (ii) Incorporation of group-contribution based thermochemistry estimation for flux analysis and pathways prediction.</td><td class="align-left"><a name="bbib0160" href="#bib0160" class="workspace-trigger">Henry et al. (2007)</a>, <a name="bbib0165" href="#bib0165" class="workspace-trigger">Henry, Broadbelt, and Hatzimanikatis (2010)</a>, <a name="bbib0210" href="#bib0210" class="workspace-trigger">Li et al. (2004)</a>, <a name="bbib0135" href="#bib0135" class="workspace-trigger">Finley et al. (2009)</a></td></tr><tr><td class="align-left">SynBioSS</td><td class="align-left">Modeling and simulation tool for synthetic biological systems. Complete enumeration of sets of biomolecular reactions based on user input of molecular parts involved in gene expression and regulation.</td><td class="align-left">(i) Multiscale simulation of the generated reaction network using stochastic algorithms.</td><td class="align-left"><a name="bbib0175" href="#bib0175" class="workspace-trigger">Hill, Tomshine, Wedding, Sotirpolous, and Kaznessis (2008)</a></td></tr><tr><td class="align-left">BioTRANS</td><td class="align-left">Computational tool for predicting metabolism of chemicals in a mixture. Generated paths of different compounds are interconnected through common metabolites. ODE models solved to predict the time profiles of the each of the compounds.</td><td class="align-left">(i) Application in modeling of biotransformations of VOCs that commonly pollute water.</td><td class="align-left"><a name="bbib0230" href="#bib0230" class="workspace-trigger">Mayeno, Yang, and Reisfeld (2005)</a></td></tr></tbody></table></div></div></div><p id="par0030">Combinatorial explosion is an important practical problem that can significantly increase execution time and lead to a large proportion of unimportant or improbable reactions. The fifth essential feature of most network generation tools, therefore, is to employ a systematic procedure to curtail this effect. When kinetic parameters are available a priori, quantitative estimates of the magnitude of the reaction rates allows for the identification of ‘important’/‘unimportant’ reactions and species that should be included in, or excluded from, the network. For example, the tool NETGEN adopts rate-based (<a name="bbib0345" href="#bib0345" class="workspace-trigger">Susnow et al., 1997</a>) network pruning criteria. This requires generation and kinetic modeling in concurrence because the rates calculated on-the-fly are used to determine if a particular species will react further. In the absence of such kinetic information, either topological or experts-based constraints can be provided. For example, species rank-based criteria (<a name="bbib0055" href="#bib0055" class="workspace-trigger">Broadbelt et al., 1995</a>) network pruning criteria prevents reactions that involve species of ranks greater than a specified value, while the tools RDL (<a name="bbib0265" href="#bib0265" class="workspace-trigger">Prickett &amp; Mavrovouniotis, 1997a</a>) and RDL++ (<a name="bbib0185" href="#bib0185" class="workspace-trigger">Hsu et al., 2008</a>) allow for the specification of constraints that molecules should satisfy to undergo a particular transformation. <a name="bbib0130" href="#bib0130" class="workspace-trigger">Faulon and Sault (2001)</a> propose stochastic (or sampling) network generation algorithms, in contrast to the deterministic scheme, for concurrent generation and reduction of networks. These algorithms scale in polynomial time but require on-the-fly estimation of rate constants which is achieved, in their case, through quantitative structure property relationships for free-radical chemistries.</p></section><section id="sec0020"><h3 class="u-h4 u-margin-m-top u-margin-xs-bottom">2.2. Topological network analysis</h3><p id="par0035">The transformations occurring on a compound to form various products can be obtained by finding: (a) a sequence of reactions that describe how the reactant gets converted into an intermediate, which could further react, and ultimately form the final product, and (b) a set of reactions that when taken together leads to an overall reaction without the net generation/consumption of any unstable reactive intermediate. The former represents “a pathway” from the initial reactant to the final product while the latter is a “mechanism” with a balanced overall stoichiometry. A <em>direct</em> mechanism is the minimal set of reactions that has zero net consumption/formation of reactive intermediates (<a name="bbib0235" href="#bib0235" class="workspace-trigger">Millner, 1964</a>); minimality ensures that no nonzero subset of this set is also a mechanism. Direct mechanisms, thus, are conceptually similar to reaction cycles. An <em>overall</em> (or complete) mechanism, further, is a set of direct mechanisms so that the reactants of the overall reaction are all initial reactants, and the desired target molecule is one of the products. Overall mechanisms, thus, are conceptually similar to a set of reaction cycles that operate simultaneously in a reaction system to convert the initial reactant to final products.</p><p id="par0040">Identifying pathways between two species in a network has wide applications in biological network analysis, and algorithms with/without tracking of individual atoms have been proposed for both weighted and unweighted networks (<a name="bbib0110" href="#bib0110" class="workspace-trigger">Croes et al., 2006</a>, <a name="bbib0155" href="#bib0155" class="workspace-trigger">Heath et al., 2010</a>). Given a list of reactions, algorithms have been proposed for identifying possible direct mechanisms using either combinatorial (<a name="bbib0225" href="#bib0225" class="workspace-trigger">Mavrovouniotis and Stephanopolous, 1992</a>, <a name="bbib0250" href="#bib0250" class="workspace-trigger">Otarod and Happel, 1992</a>) or graph theoretic methods (<a name="bbib0125" href="#bib0125" class="workspace-trigger">Fan et al., 2002</a>, <a name="bbib0220" href="#bib0220" class="workspace-trigger">Lin et al., 2009</a>).</p><p id="par0045">A concept closely related to that of mechanisms is the linear-algebra-based flux analysis wherein all possible independent basis set of solution vectors that solve the stoichiometric steady-state mass balance equations are identified (<a name="bbib0255" href="#bib0255" class="workspace-trigger">Papin et al., 2004</a>). Two commonly applied flux analysis techniques in systems biology are extreme pathways (<a name="bbib0325" href="#bib0325" class="workspace-trigger">Schilling, Letscher, &amp; Palsson, 2000</a>) and elementary flux modes (<a name="bbib0335" href="#bib0335" class="workspace-trigger">Schuster et al., 1994</a>).</p></section></section><section id="sec0025"><h2 class="u-h3 u-margin-l-top u-margin-xs-bottom">3. The structure of RING: an overview</h2><div><p id="par0050">RING builds on and extends the network generators listed in <a name="btbl0005" href="#tbl0005" class="workspace-trigger">Table 1</a> in two major aspects. First, RING provides a generic scope by allowing for representing intermediates and describing reaction rules applicable to most organic chemical transformations. Second, RING allows for topological network analysis through identification of pathways and mechanisms in the generated network. <a name="bfig0005" href="#fig0005" class="workspace-trigger">Fig. 1</a> shows the input–output structure of RING as well as the three internal modules – a language compiler, a reaction network generator, and a post-processing module.</p><figure class="figure text-xs" id="fig0005"><span><img src="https://ars.els-cdn.com/content/image/1-s2.0-S0098135412001846-gr1.jpg" height="116" alt=""><ol class="links-for-figure"><li><a class="anchor download-link u-font-sans" href="https://ars.els-cdn.com/content/image/1-s2.0-S0098135412001846-gr1.jpg" target="_blank" download="" title="Download full-size image"><span class="anchor-text">Download : <span class="download-link-title">Download full-size image</span></span></a></li></ol></span><span class="captions"><span><p id="spar0015"><span class="label">Fig. 1</span>. The modular structure of RING. Dashed arrows show internal information flow, solid arrows refer to inputs and outputs.</p></span></span></figure></div><section id="sec0030"><h3 class="u-h4 u-margin-m-top u-margin-xs-bottom">3.1. Reaction language compiler</h3><p id="par0055">The compiler acts as an interface between the user and network generation and post-processing components of RING by translating the inputs into relevant instructions used internally by the network generator. The inputs into RING are written as a program in an English-like reaction language and include the initial reactants, reaction rules, and set of post-processing instructions.</p><p id="par0060">The utility of a language interface in network generation was demonstrated in RDL (<a name="bbib0265" href="#bib0265" class="workspace-trigger">Prickett &amp; Mavrovouniotis, 1997a</a>) and RDL++ (<a name="bbib0185" href="#bib0185" class="workspace-trigger">Hsu et al., 2008</a>). The syntax of the language closely resembles chemistry parlance so that debugging a program written in the language essentially involves proofreading. RING's language interface has been developed using SILVER, an attribute grammar specification language (<a name="bbib0365" href="#bib0365" class="workspace-trigger">Van Wyk, Bodin, Gao, &amp; Krishnan, 2010</a>). In addition to providing an interface like in RDL++, the language, as discussed in detail in <a name="bbib0295" href="#bib0295" class="workspace-trigger">Rangarajan, Bhan, and Daoutidis (in press)</a>, can: (a) catch chemistry-specific inconsistencies as well as syntactical errors, (b) improve speed of execution through domain-specific optimization of input instructions, and (c) allow for independent extensions in terms of syntax and semantics to the original grammar.</p></section><section id="sec0035"><h3 class="u-h4 u-margin-m-top u-margin-xs-bottom">3.2. Reaction network generator</h3><p id="par0065">The information on initial reactants and the translated reaction rules is used by the network generator to construct the reaction network by iterative application of the rules upon the initial reactants and the products generated thereof. The output from the network generator is a list of all possible species and reactions consistent with the rules specified. Several Cheminformatics algorithms to manipulate the chemical information of a molecule, extract topological information of molecules, and generate reactions have been adopted in RING. Further, standard Cheminformatics formats to input, represent, store, and retrieve the chemical information pertaining to molecules, such as elemental composition, valency, charge, bonding, and electronic information have also been used.</p></section><section id="sec0040"><h3 class="u-h4 u-margin-m-top u-margin-xs-bottom">3.3. Post-processing module</h3><p id="par0070">Specific information regarding the constituents of/transformations within the network can be “mined” from the generated reaction network using the several post-processing options available in RING. For example, the generated reaction network may be large enough to preclude manual analysis to obtain specific information such as whether and how a particular molecule (or a class of molecules) is (are) formed. Such queries can be input in the form of instructions, or “rules”, specifying the analysis that is sought.</p><section id="sec0045"><h4 class="u-margin-m-top u-margin-xs-bottom">3.3.1. Lumping</h4><p id="par0075">Complex reaction networks can be composed of several species that are structural isomers of each other. If these similar species are grouped together into lumps, the total network size can be reduced significantly. For example, <a name="bbib0185" href="#bib0185" class="workspace-trigger">Hsu et al. (2008)</a> showed for acid-catalyzed aromatization of propane that the number of reactions without lumping structural isomers was more than a million while structural lumping resulted in only 605 distinct reactions. Analytical identification of lumps of molecules, instead of individual species, is a more feasible option when the number of species in a system is large (<a name="bbib0285" href="#bib0285" class="workspace-trigger">Quann &amp; Jaffe, 1996</a>) because of experimental difficulties in distinguishing between isomers. Furthermore, lumping can lead to a reduced network that can be used to formulate a simplified lower-dimensional mathematical model that is easier to solve. Lumping strategies have been proposed in hydrocarbon processing wherein, a lump is represented either by a single likely structure and reactions are written on that basis (<a name="bbib0280" href="#bib0280" class="workspace-trigger">Quann &amp; Jaffe, 1992</a>), or as a mixture with an estimated internal distribution and an overall reaction computed on the basis of composition and kinetics information (<a name="bbib0300" href="#bib0300" class="workspace-trigger">Ranzi, Dente, Goldaniga, Bozzano, &amp; Faravelli, 2001</a>). In contrast to these methods, lumping methods for dimension reduction in kinetic models through mathematical transformations have also been proposed (<a name="bbib0205" href="#bib0205" class="workspace-trigger">Kuo and Wei, 1969</a>, <a name="bbib0215" href="#bib0215" class="workspace-trigger">Li and Rabitz, 1989</a>, a brief review in <a name="bbib0180" href="#bib0180" class="workspace-trigger">Ho, 2008</a>).</p><p id="par0080">Lumping instructions, as an input feature is not available in the tools listed in <a name="btbl0005" href="#tbl0005" class="workspace-trigger">Table 1</a>, though RDL++ (<a name="bbib0185" href="#bib0185" class="workspace-trigger">Hsu et al., 2008</a>) can group together hydrocarbons in a post-processing step. In RING, however, a generic lumping scheme, that identifies functional equivalence between isomers, has been implemented. In general, two molecules can be lumped only when there is a one-to-one mapping between their atoms; mapping in this case exists when the two atoms belong to the same functional group. Such functional equivalence cannot be inferred from molecular formula and structure alone for non-hydrocarbon molecules such as oxygenates. We have, therefore, developed a new lumping algorithm that takes into consideration the environment of each atom in a molecule to decide the lump it belongs to. For example, ethanol (CH<sub>3</sub>CH<sub>2</sub>OH) and dimethyl ether (CH<sub>3</sub>OCH<sub>3</sub>) both have the same molecular formula; however, they have different functional groups. Ethanol has a C–OH group that the ether does not possess; hence, there is no one-to-one mapping between the oxygen atom of ethanol and the ether. Therefore, the two molecules are not lumped. On the other hand, 2-pentanol and 3-pentanol are both functionally equivalent (<a name="bsec0160" href="#sec0160" class="workspace-trigger">S2 in Supplementary material</a>) are lumped together as secondary five-carbon alcohols. 1-Pentanol, being a primary alcohol, will naturally not be a constituent of this lump.</p></section><section id="sec0050"><h4 class="u-margin-m-top u-margin-xs-bottom">3.3.2. Pathway identification</h4><p id="par0085">RING can identify pathways to user-specified molecules from initial reactants that satisfy user-specified constraints on the nature of pathways (maximum length, presence/absence of reactions in the pathway, etc.). A reverse depth-first network traversal algorithm (<a name="bbib0295" href="#bib0295" class="workspace-trigger">Rangarajan et al., in press</a>) is used to identify all possible pathways exhaustively.</p></section><section id="sec0055"><h4 class="u-margin-m-top u-margin-xs-bottom">3.3.3. Mechanism identification</h4><p id="par0090">In RING, all overall mechanisms from the initial reactants to user-specified products can be identified. Further, all direct mechanisms forming any specified molecule can also be found. The algorithms of <a name="bbib0225" href="#bib0225" class="workspace-trigger">Mavrovouniotis and Stephanopoulos (1992)</a> and <a name="bbib0125" href="#bib0125" class="workspace-trigger">Fan et al. (2002)</a> identify all possible direct mechanisms from a given list of reactions. Since mechanisms are only sought for specified products, these two algorithms cannot be directly used within RING. An adapted version of the pathways algorithm is instead used that seeks only the relevant direct or overall mechanisms.</p></section></section></section><section id="sec0060"><h2 class="u-h3 u-margin-l-top u-margin-xs-bottom">4. RING: inputs and outputs</h2><div><p id="par0095">Inputs into RING are: (a) initial reactants, (b) global constraints, (c) reaction rules, and (d) post-processing instructions describing the desired analyses of the network. The outputs from RING are lists of: (a) species, (b) reactions, (c) lumps and their constituents, (d) pathways to specified products, (e) mechanisms to specified products, and (f) species or reactions extracted from the network from a molecule or reaction query. In this section, the inputs into and outputs from RING are discussed in the context of an example – dehydration of fructose to form HMF. The system is typically carried out in aqueous phase and is catalyzed by mineral acids (<a name="bbib0005" href="#bib0005" class="workspace-trigger">Antal, Mok, &amp; Richards, 1990</a>). An illustrative program for this system is shown in <a name="btbl0010" href="#tbl0010" class="workspace-trigger">Table 2</a>.</p><div class="tables frame-topbot rowsep-0 colsep-0" id="tbl0010"><span class="captions"><span><p id="spar0060"><span class="label">Table 2</span>. Sample input code for Fructose-to-HMF system.</p></span></span><div class="groups"><table><thead class="valign-top"><tr><th scope="col" class="rowsep-1 align-left">Reaction rules</th><th scope="col" class="rowsep-1 align-left">Post-processing</th></tr></thead><tbody><tr><td class="align-left">1. input reactant “O1C(CO)(O)C(O)C(O)C1(CO)”</td><td class="align-left">27. lump all isomers{</td></tr><tr><td class="align-left">2. input reactant “[H+]”</td><td class="align-left">28. represent acyclic with farthest apart</td></tr><tr><td class="align-left">3. //Defining global constraints</td><td class="align-left">29. represent cyclic with farthest apart}</td></tr><tr><td class="align-left">4. global constraints on Molecule f</td><td class="align-left">30. find pathways to mol{</td></tr><tr><td class="align-left">5. fragment a{</td><td class="align-left">31. mol is “OCc1oc(C<img src="https://sdfestaticassets-us-east-1.sciencedirectassets.com/shared-assets/55/entities/dbnd.gif" alt="double bond" style="vertical-align: middle;">O)cc1”</td></tr><tr><td class="align-left">6. C+ labeled 1</td><td class="align-left">32. } constraints {</td></tr><tr><td class="align-left">7. any atom labeled 2 double bond to 1}</td><td class="align-left">33. maximum length 11</td></tr><tr><td class="align-left">8. ! Molecule contains a</td><td class="align-left">34. contains &lt;= 2 rule Hshift</td></tr><tr><td class="align-left">9. Molecule.size &lt; 15 &amp;&amp; Molecule.charge &lt; 2</td><td class="align-left">35. eliminate similar pathways</td></tr><tr><td class="align-left">10. fragment b {</td><td class="align-left">36. } store in “HMFPathways.txt”</td></tr><tr><td class="align-left">11. C labeled c1</td><td class="align-left">37. find complete mechanisms to mol{</td></tr><tr><td class="align-left">12. C labeled c2 double bond to c1</td><td class="align-left">38. mol is “OCc1oc(C<img src="https://sdfestaticassets-us-east-1.sciencedirectassets.com/shared-assets/55/entities/dbnd.gif" alt="double bond" style="vertical-align: middle;">O)cc1”} overall constraints{</td></tr><tr><td class="align-left">13. X labeled x1 double bond to c2}</td><td class="align-left">39. maximum length 12</td></tr><tr><td class="align-left">14. ! Molecule contains b}</td><td class="align-left">40. maximum cycles 4} cycles constraints {</td></tr><tr><td class="align-left">15. //alcohol protonation rule</td><td class="align-left">41. maximum length 4</td></tr><tr><td class="align-left">16. rule alcoholprot{</td><td class="align-left">42. contains &lt;= 2 rule Hshift</td></tr><tr><td class="align-left">17. neutral reactant r1{</td><td class="align-left">43. eliminate similar mechanisms} store in “HMFMechanisms.txt”</td></tr><tr><td class="align-left">18. C labeled c1 {! connected to &gt; 1 O with any bond}</td><td class="align-left">44. find mol{</td></tr><tr><td class="align-left">19. nonringatom O labeled o1 single bond to c1}</td><td class="align-left">45. mol is cyclic &amp;&amp; mol.size &gt; 5 &amp;&amp; mol is neutral</td></tr><tr><td class="align-left">20. positive reactant proton{</td><td class="align-left">46. } store in “CyclicNeutralMolecules.txt”</td></tr><tr><td class="align-left">21. H+ labeled h1}</td><td class="align-left">47. find reactions{</td></tr><tr><td class="align-left">22. constraints {r1.size &lt; 15 &amp;&amp; r1.size &gt; 2}</td><td class="align-left">48. rule is only alcoholprot</td></tr><tr><td class="align-left">23. form bond (o1, h1)</td><td class="align-left">49. reaction with 1 reactant mol{mol is cyclic}</td></tr><tr><td class="align-left">24. modify atomtype (o1, O+)</td><td class="align-left">50. } store in “CyclicAlcProtRxns.txt”</td></tr><tr><td class="align-left">25. modify atomtype (h1, H)}</td><td></td></tr><tr><td class="align-left">26. //More reaction rules here</td><td></td></tr></tbody></table></div></div></div><section id="sec0065"><h3 class="u-h4 u-margin-m-top u-margin-xs-bottom">4.1. Initial reactants</h3><div><p id="par0100">Initial reactants are represented as SMILES-like strings (<a name="bbib0380" href="#bib0380" class="workspace-trigger">Weininger, 1988</a>). Lines 1 and 2 in <a name="btbl0010" href="#tbl0010" class="workspace-trigger">Table 2</a> describe the reactants for the Fructose-to-HMF system (Fructofuranose and proton to represent the acid), written in SMILES. A SMILES parser has been developed for RING which recognizes the elements C, H, N, O, S, and P. On many occasions, inorganic atoms such as Al, Cu, Pt, etc. are involved in chemical transformations, for example, as active sites (catalysts, ligands) or as reagents (electrophiles, nucleophiles, etc.) in substitution reactions, and have to be represented and manipulated during reaction generation. To this end, a new feature, <em>composite atoms</em>, has been added to represent a miscellaneous group of atoms as a single atom for convenience in representation and manipulation. A composite atom, within RING, is a user-defined entity used to represent: (a) an atom other than the recognized elements or (b) a group of atoms considered together as a “unified atom”. These atoms are represented as alphanumeric strings enclosed within curly braces, for example, Platinum atoms can be represented as ‘{Pt}’. The name given to a composite atom is arbitrary and does not have to be an element name. For example, two different types of active sites (AS) on a catalyst can be represented as {AS1} and {AS2}. These atoms have to be defined using “define composite atom” statement. Once defined, RING recognizes the composite atom during reaction rule specification and reaction generation. Examples employing composite atoms are discussed in the second part of this series (<a name="bbib0295" href="#bib0295" class="workspace-trigger">Rangarajan et al., in press</a>). A second feature that has been added into SMILES in RING is the definition of non-bonded interactions – partial and hydrogen bond. This interaction is represented using “_” and can be used to depict, for example, dimeric alcohol intermediates in zeolites (<a name="bbib0080" href="#bib0080" class="workspace-trigger">Chiang &amp; Bhan, 2010</a>) co-adsorbed on a zeolite, or adsorption of molecules on a metal surface as shown in <a name="bfig0010" href="#fig0010" class="workspace-trigger">Fig. 2</a>. The use of SMILES, with these two additional features, allow for rigorous description of the intermediates involved in different types of chemistries.</p><figure class="figure text-xs" id="fig0010"><span><img src="https://ars.els-cdn.com/content/image/1-s2.0-S0098135412001846-gr2.jpg" height="197" alt=""><ol class="links-for-figure"><li><a class="anchor download-link u-font-sans" href="https://ars.els-cdn.com/content/image/1-s2.0-S0098135412001846-gr2.jpg" target="_blank" download="" title="Download full-size image"><span class="anchor-text">Download : <span class="download-link-title">Download full-size image</span></span></a></li></ol></span><span class="captions"><span><p id="spar0020"><span class="label">Fig. 2</span>. SMILES strings with composite atoms and non-bonded interactions: (i) dimeric intermediate in DEE synthesis from ethanol on proton form zeolite catalysts (<a name="bbib0080" href="#bib0080" class="workspace-trigger">Chiang &amp; Bhan, 2010</a>); (ii) surface species formed upon dissociative adsorption of ethanol.</p></span></span></figure></div></section><section id="sec0070"><h3 class="u-h4 u-margin-m-top u-margin-xs-bottom">4.2. Reaction rules</h3><p id="par0105">These describe the chemical transformations that take place in the reaction system (<a name="bbib0290" href="#bib0290" class="workspace-trigger">Rangarajan et al., 2010</a>) and comprise of: (i) the set of atoms that participate in the reaction, called the reaction center or reactant patterns, (ii) the final electronic configuration of the atoms in the reaction center and the bonding between them, and (iii) constraints that govern the reaction center and the entire molecule. These constraints have to be satisfied by the reactant molecule prior to the application of the reaction rules. In addition, constraints can also be imposed on the products formed and are classified under molecular constraints.</p><div><p id="par0110">Each reaction rule defines a particular chemical transformation which can be elementary or non-elementary, and be unimolecular or bimolecular. For a given reaction system, the reaction rules can be deduced from the literature and from knowledge of the operating conditions. For example, gas phase pyrolysis of hydrocarbons or oxygenates typically involves homolytic bond cleavage/formation, β-scission, hydride shifts, etc., while heterogeneous solid acid catalysis would involve adsorption/desorption of hydrocarbons, β-scission, alkylation, etc. RING allows for the specification of an additional reactant identical to one of the reactant(s). This feature has been implemented mainly to address elementary steps on surfaces where two adjacent identical surface species (or sites) are involved, such as desorption of surface hydrogen species on palladium to form molecular hydrogen (<a name="bbib0090" href="#bib0090" class="workspace-trigger">Conrad, Ertl, &amp; Latta, 1974</a>). This duplication allows for an exception whereby termolecular reactions can de described when the third reactant pattern is identical to one of the other two, such as in the case of dissociative adsorption of molecular hydrogen (the reverse of the desorption step) shown in <a name="bfig0015" href="#fig0015" class="workspace-trigger">Fig. 3</a>. Lines 16–25 in <a name="btbl0010" href="#tbl0010" class="workspace-trigger">Table 2</a> describe the reaction rule for protonation of a hydroxy group, shown in <a name="bfig0020" href="#fig0020" class="workspace-trigger">Fig. 4</a>. The rule essentially describes the abstraction of a proton by the oxygen atom of an alcohol to form an oxonium ion.</p><figure class="figure text-xs" id="fig0015"><span><img src="https://ars.els-cdn.com/content/image/1-s2.0-S0098135412001846-gr3.jpg" height="87" alt=""><ol class="links-for-figure"><li><a class="anchor download-link u-font-sans" href="https://ars.els-cdn.com/content/image/1-s2.0-S0098135412001846-gr3.jpg" target="_blank" download="" title="Download full-size image"><span class="anchor-text">Download : <span class="download-link-title">Download full-size image</span></span></a></li></ol></span><span class="captions"><span><p id="spar0025"><span class="label">Fig. 3</span>. Dissociative adsorption of molecular hydrogen on metal sites (Pd).</p></span></span></figure><figure class="figure text-xs" id="fig0020"><span><img src="https://ars.els-cdn.com/content/image/1-s2.0-S0098135412001846-gr4.jpg" height="39" alt=""><ol class="links-for-figure"><li><a class="anchor download-link u-font-sans" href="https://ars.els-cdn.com/content/image/1-s2.0-S0098135412001846-gr4.jpg" target="_blank" download="" title="Download full-size image"><span class="anchor-text">Download : <span class="download-link-title">Download full-size image</span></span></a></li></ol></span><span class="captions"><span><p id="spar0030"><span class="label">Fig. 4</span>. Sample input reaction rule in RING – elementary step involving protonation of an alcohol.</p></span></span></figure></div><section id="sec0075"><h4 class="u-margin-m-top u-margin-xs-bottom">4.2.1. Reactant patterns</h4><p id="par0115">Reaction centers (patterns) are written as shown in lines 17–21 in <a name="btbl0010" href="#tbl0010" class="workspace-trigger">Table 2</a>. The hydroxy pattern is described in lines 17–19 – a carbon atom, labeled ‘c1’, that is single bonded to an oxygen atom not in a ring, labeled ‘o1’. Lines 20–21 describe a proton – a positive hydrogen atom labeled ‘h1’. All atoms in a reactant pattern are labeled for unambiguous identification while describing transformations. Each reactant pattern describes one of the reactants. Rules can be unimolecular (single pattern), bimolecular (two patterns), ternary (consisting of two identical and a second distinct pattern), and intramolecular (special case of bimolecular rules). Therefore, bimolecular reactions have two reactant patterns which together constitute the reaction center. Ternary reaction rules can also be specified when two of the three patterns are identical. Bimolecular reaction rules can be set to also/only allow intramolecular reactions. In such a case, an intramolecular reaction will occur only if both the patterns exist in a single reactant.</p></section><section id="sec0080"><h4 class="u-margin-m-top u-margin-xs-bottom">4.2.2. Constraints specification</h4><p id="par0120">Constraints, as discussed in Section <a name="bsec0010" href="#sec0010" class="workspace-trigger">2</a>, are essential in preventing a combinatorial explosion in the size of the generated network. In RING, two types of constraints can be specified: molecular and atom environment constraints. The atom environment constraints, given along with reactant pattern specification, describe the spatial properties of the atoms in the reaction center. Specifically, they describe: (a) what the neighboring atoms and fragments are, and (b) topological characteristics of the atoms – ring atom, aromatic atom or allylic atom, largest size of the ring in which atom is present, etc. The molecular constraints are constraints on (a) charge, (b) size, (c) presence (or absence) of structural features, and (d) shape and presence (or absence) of topological characteristics. Atom constraints are specified as shown in line 18 of <a name="btbl0010" href="#tbl0010" class="workspace-trigger">Table 2</a>, wherein, the constraint restricts the carbon ‘c1’ to be connected to one or fewer oxygen atoms. This ensures that the carbon is not part of an ester or an acid functional group.</p><p id="par0125">Molecular constraints are specified as shown in line 22 of <a name="btbl0010" href="#tbl0010" class="workspace-trigger">Table 2</a>. The carbonyl species has an upper and lower bound on the size (defined as the number of non-hydrogen atoms in it). Lines 17 and 20 also impose charge constraints on the reactants by using the prefixes “neutral” and “positive”. Combined constraints involving both reactants and constraints on products can also be specified. This constraint specification scheme, inclusive of reactant and product molecular constraints and reaction center atom constraints, thus, provides the user with the capability of imposing restrictions to varying levels of severity.</p></section><section id="sec0085"><h4 class="u-margin-m-top u-margin-xs-bottom">4.2.3. Transformations</h4><p id="par0130">A reaction rule is complete when the final configuration of atoms and bonds of the reaction center is specified. Transformations in RING, similar to those in RDL (<a name="bbib0265" href="#bib0265" class="workspace-trigger">Prickett &amp; Mavrovouniotis, 1997a</a>) and RDL++ (<a name="bbib0185" href="#bib0185" class="workspace-trigger">Hsu et al., 2008</a>) involve either changes in connectivity such as bond formation/cleavage and modification of the bond order, or changes in electronic configuration such as adding or removing charge, lone pairs, and unpaired electrons. The changes in electronic configuration are described by atomtype modifications – an atom of one atomtype changes to another atomtype (e.g., C forming C+). Lines 23–25 of <a name="btbl0010" href="#tbl0010" class="workspace-trigger">Table 2</a> describe the transformations for the case of alcohol protonation: (a) formation of a single bond between oxygen atom, ‘o1’, and hydrogen atom, ‘h1’, (b) formation of an oxonium ion, and (c) loss of charge on the proton to form neutral hydrogen. The description of changes in the bond connectivity and electronic configurations, along with reactant pattern description, provides complete information on the initial and final state of each atom in the reaction center. This allows for: (a) catching incorrect rules that violate valency constraints of atoms directly by the language compiler prior to network generation, and (b) correctly identifying changes in topological characteristics of molecules such as in steps that break aromaticity. Such a description of transformations is independent of the type of chemistry and hence applicable to most organic chemical transformations – homogeneous/heterogeneous catalytic/non-catalytic chemistries.</p><p id="par0135">The Fructose-to-HMF system has several additional rules: protonation of carbonyl and C<img src="https://sdfestaticassets-us-east-1.sciencedirectassets.com/shared-assets/55/entities/dbnd.gif" alt="double bond" style="vertical-align: middle;">C groups (and their reverse steps), hydride shift, allyl shift, and dehydration of oxonium ion. <a name="bsec0160" href="#sec0160" class="workspace-trigger">S1 in Supporting material</a> contains the corresponding reaction language code for these elementary steps. These rules, while only a subset of all acid-catalyzed elementary steps, form the likely steps for aqueous phase mineral acid catalyzed system at moderate temperatures (∼100&nbsp;°C).</p></section></section><section id="sec0090"><h3 class="u-h4 u-margin-m-top u-margin-xs-bottom">4.3. Global constraints</h3><p id="par0140">These are molecular constraints on the entire reaction system applicable to all molecules in the network. For example, global constraints can fix the maximum possible size of a molecule in a system. For the Fructose-to-HMF system, lines 4–14 describe molecular constraints applicable to all molecules at all times – the maximum size of a molecule is 15 atoms (line 9), the maximum allowed charge on a molecule is 1 (line 9), a positively charged atom can never be connected to any atom with a double bond (lines 5–8), and there are no consecutive double bonds (lines 10–14). It can be noted that the two fragments, ‘a’ and ‘b’, are defined in a manner similar to the reactant pattern definition, and refer to specific patterns in the molecule that should or should not be present in the molecule. Such fragment constraints can also be imposed in reaction rules.</p></section><section id="sec0095"><h3 class="u-h4 u-margin-m-top u-margin-xs-bottom">4.4. Post-processing instructions</h3><section id="sec0100"><h4 class="u-margin-m-top u-margin-xs-bottom">4.4.1. Lumping strategy</h4><p id="par0145">In RING, the user can specify: (a) if lumping based on functional equivalence is required, (b) if the representative molecule of the lump is the isomer that has branches farthest/closest apart, and (c) if further lumping of hydrocarbons, such as paraffins, olefins, naphthenes, and aromatics based on size is required. While line 27 specifies that lumping is required, lines 28–29 in <a name="btbl0010" href="#tbl0010" class="workspace-trigger">Table 2</a> describe how acyclic and cyclic molecules satisfying the condition for lumping, are both represented by the constituent that has its branches farthest apart. For example, in the lump of five-carbon secondary alcohols (Section <a name="bsec0045" href="#sec0045" class="workspace-trigger">3.3.1</a>), the lump representative, according to line 28, would be 3-pentanol because its branch (the C–OH group) is farther apart from both the ends when compared to 2-pentanol, the other constituent.</p></section><section id="sec0105"><h4 class="u-margin-m-top u-margin-xs-bottom">4.4.2. Pathways identification</h4><p id="par0150">Pathway querying, in RING, consists of two steps – a description of the target molecule(s) and the nature of the pathways desired. Target molecules are described by molecular constraints, while constraints on the nature of pathways – pathway constraints – are described subsequently to put a limit on path length, describe presence/absence of rules and molecules, and to describe constraints on the reactants and products in one or more reactions of the pathway. Lines 30–36 in <a name="btbl0010" href="#tbl0010" class="workspace-trigger">Table 2</a> show a pathway query to HMF, “OCc1oc(C<img src="https://sdfestaticassets-us-east-1.sciencedirectassets.com/shared-assets/55/entities/dbnd.gif" alt="double bond" style="vertical-align: middle;">O)cc1”. The constraints on the pathway include a maximum length constraint of 11 reactions in the sequence, a limit on the number of occurrences of a reaction corresponding to the rule Hshift (hydride shift, see <a name="bsec0160" href="#sec0160" class="workspace-trigger">Supporting information</a>), and a requirement that similar pathways<a name="bfn0005" href="#fn0005" class="workspace-trigger"><sup>1</sup></a> be eliminated.</p></section><section id="sec0110"><h4 class="u-margin-m-top u-margin-xs-bottom">4.4.3. Mechanism identification</h4><p id="par0155">Lines 37–43 in <a name="btbl0010" href="#tbl0010" class="workspace-trigger">Table 2</a> describe an overall mechanism query to find mechanisms to HMF. The mechanisms can have several cycles; therefore, constraints can be imposed on the overall mechanism (lines 39–40) as well as on individual cycles (lines 41–43). The overall constraints restrict the total number of reactions and cycles in the complete mechanism. The individual cycle constraints restrict the number of reactions in each cycle, and limit the number of occurrences of a reaction corresponding to the rule Hshift. Further, the requirement “eliminate similar mechanisms” is imposed so that multiple cycles having the same overall reaction are not obtained.</p></section><section id="sec0115"><h4 class="u-margin-m-top u-margin-xs-bottom">4.4.4. Molecule and reaction queries</h4><p id="par0160">The generated network can be queried for specific components of the network – species and reactions. The molecule queries are given as shown in lines 44–46 in <a name="btbl0010" href="#tbl0010" class="workspace-trigger">Table 2</a> wherein all molecules satisfying the specified molecular constraints – cyclic and neutral – are sought. Further, reaction queries seeking all reactions of the network satisfying specific types of reactions, reactants, or products, can also be input. Lines 47–50 in <a name="btbl0010" href="#tbl0010" class="workspace-trigger">Table 2</a> show an example of reaction query for obtaining all alcohol protonation reactions involving a cyclic reactant.</p></section></section><section id="sec0120"><h3 class="u-h4 u-margin-m-top u-margin-xs-bottom">4.5. Outputs from RING</h3><section id="sec0125"><h4 class="u-margin-m-top u-margin-xs-bottom">4.5.1. Reaction network</h4><div><p id="par0165">The reaction network is a list of species and a list of reactions, both written in SMILES format. The format for reactions essentially involves concatenating SMILES strings of reactants and products of that reaction with appropriate delimiters (‘.’ demarcates two different reactants/products, while ‘&gt;&gt;’ separates reactants from products). A SMILES generator has been developed that constructs the individual strings for molecules from their internal representation and appends them together to get the reaction string. <a name="bfig0025" href="#fig0025" class="workspace-trigger">Fig. 5</a> shows a sample set of reactions constituting the network, in both SMILES and graphical form. Several chemistry packages parse SMILES strings to generate a graphical output of molecules and reactions; the reactions in <a name="bfig0025" href="#fig0025" class="workspace-trigger">Fig. 5</a> were generated using ChemDraw (<a name="bbib0070" href="#bib0070" class="workspace-trigger">CambridgeSoft, 2011</a>).</p><figure class="figure text-xs" id="fig0025"><span><img src="https://ars.els-cdn.com/content/image/1-s2.0-S0098135412001846-gr5.jpg" height="290" alt=""><ol class="links-for-figure"><li><a class="anchor download-link u-font-sans" href="https://ars.els-cdn.com/content/image/1-s2.0-S0098135412001846-gr5.jpg" target="_blank" download="" title="Download full-size image"><span class="anchor-text">Download : <span class="download-link-title">Download full-size image</span></span></a></li></ol></span><span class="captions"><span><p id="spar0035"><span class="label">Fig. 5</span>. Sample reactions of the Fructose-to-HMF network generated by RING.</p></span></span></figure></div></section><section id="sec0130"><h4 class="u-margin-m-top u-margin-xs-bottom">4.5.2. Lumping results</h4><div><p id="par0170">Results from a lumping analysis are output as a list of lumps wherein each lump is represented by its representative molecule (identified based on user-specified criterion) and is followed by a list of species belonging to the lump. <a name="bfig0030" href="#fig0030" class="workspace-trigger">Fig. 6</a> shows, pictorially, one of the lumps identified by RING – a lump of Fructose derivatives containing three hydroxy groups, an oxonium ion, a methyl group, and a carbon–carbon double bond. The eight molecules shown in the figure are lumped into one lump represented by the lump representative shown. Once the molecules are lumped, each reaction can be rewritten in terms of lumps to get a lumped reaction. Since the number of lumps is lesser than the number of species in a network (about 550 species and 240 lumps), the number of lumped reactions is also fewer (about 650 lumped reactions from a network of about 1200 reactions). Lumping, thus, reduces the network size by a factor of 2. For some systems, such as propane aromatization, as shown in the accompanying paper, lumping is shown to lead to reduction in network size by several orders of magnitude (<a name="bbib0295" href="#bib0295" class="workspace-trigger">Rangarajan et al., in press</a>).</p><figure class="figure text-xs" id="fig0030"><span><img src="https://ars.els-cdn.com/content/image/1-s2.0-S0098135412001846-gr6.jpg" height="380" alt=""><ol class="links-for-figure"><li><a class="anchor download-link u-font-sans" href="https://ars.els-cdn.com/content/image/1-s2.0-S0098135412001846-gr6.jpg" target="_blank" download="" title="Download full-size image"><span class="anchor-text">Download : <span class="download-link-title">Download full-size image</span></span></a></li></ol></span><span class="captions"><span><p id="spar0040"><span class="label">Fig. 6</span>. Sample lump and its representative in the Fructose-to-HMF reaction network identified by RING.</p></span></span></figure></div></section><section id="sec0135"><h4 class="u-margin-m-top u-margin-xs-bottom">4.5.3. Pathway identification results</h4><div><p id="par0175">Pathways are obtained as a list of reaction sequences wherein each sequence is, in itself, a list of reactions that traces the product formation. <a name="bfig0035" href="#fig0035" class="workspace-trigger">Fig. 7</a> depicts one of the pathways for the conversion of fructose to HMF as a sequence of reactions. The pathway involves three alcohol-protonation and dehydration steps, while one of the steps is essentially an enol–carbonyl tautomerism step. <a name="bbib0005" href="#bib0005" class="workspace-trigger">Antal et al. (1990)</a> propose this sequence of steps as the dominant route for HMF synthesis from fructose in aqueous solutions catalyzed by mineral acids.</p><figure class="figure text-xs" id="fig0035"><span><img src="https://ars.els-cdn.com/content/image/1-s2.0-S0098135412001846-gr7.jpg" height="476" alt=""><ol class="links-for-figure"><li><a class="anchor download-link u-font-sans" href="https://ars.els-cdn.com/content/image/1-s2.0-S0098135412001846-gr7.jpg" target="_blank" download="" title="Download full-size image"><span class="anchor-text">Download : <span class="download-link-title">Download full-size image</span></span></a></li></ol></span><span class="captions"><span><p id="spar0045"><span class="label">Fig. 7</span>. One pathway from fructose to HMF in the Fructose-to-HMF reaction network identified by RING. The sequential representation of reaction is an adaptation of the reaction list that RING generates.</p></span></span></figure></div></section><section id="sec0140"><h4 class="u-margin-m-top u-margin-xs-bottom">4.5.4. Mechanism identification results</h4><div><p id="par0180"><a name="bfig0040" href="#fig0040" class="workspace-trigger">Fig. 8</a> shows the overall (complete) mechanism that corresponds to the pathway described above. The mechanism has four catalytic cycles (<a name="bfig0040" href="#fig0040" class="workspace-trigger">Fig. 8</a>) – three dehydration cycles and one isomerization cycle. The overall reaction, thus, is a balanced reaction with one molecule of HMF and three molecules of water formed from each fructose molecule. We can further query to obtain the constituent elementary step reactions of these individual cycles through identifying direct mechanisms.</p><figure class="figure text-xs" id="fig0040"><span><img src="https://ars.els-cdn.com/content/image/1-s2.0-S0098135412001846-gr8.jpg" height="433" alt=""><ol class="links-for-figure"><li><a class="anchor download-link u-font-sans" href="https://ars.els-cdn.com/content/image/1-s2.0-S0098135412001846-gr8.jpg" target="_blank" download="" title="Download full-size image"><span class="anchor-text">Download : <span class="download-link-title">Download full-size image</span></span></a></li></ol></span><span class="captions"><span><p id="spar0050"><span class="label">Fig. 8</span>. Example of one complete mechanism having four catalytic cycles for HMF production from fructose, identified by RING.</p></span></span></figure></div></section></section></section><section id="sec0145"><h2 class="u-h3 u-margin-l-top u-margin-xs-bottom">5. Discussion</h2><p id="par0185">The input options and output features of RING allow for systematic rule-based elucidation of complex reaction networks through: (a) construction of the reaction network from reaction rules, and (b) analysis of the transformations in the network through post-processing instructions (or rules). Such analysis is particularly relevant in biomass conversion as upgrading oxygenates to produce fuels and chemicals involves multiple thermochemical steps to remove oxygen atoms present in different functional groups (<a name="bbib0095" href="#bib0095" class="workspace-trigger">Corma et al., 2007</a>, <a name="bbib0190" href="#bib0190" class="workspace-trigger">Huber et al., 2005</a>, <a name="bbib0330" href="#bib0330" class="workspace-trigger">Schmidt and Dauehauer, 2007</a>). RING currently does not take into consideration the kinetics of each of the steps and hence cannot apply features such as rate-based pruning (<a name="bbib0345" href="#bib0345" class="workspace-trigger">Susnow et al., 1997</a>).</p><p id="par0190">The construction and analysis features of RING can be applied for analyzing complex reaction networks in three distinct classes of problems. We discuss each of these through illustrative examples in the second part of this series (<a name="bbib0295" href="#bib0295" class="workspace-trigger">Rangarajan et al., in press</a>):<dl class="list"><dt class="list-label">(a)</dt><dd class="list-description"><p id="par0195">If the elementary chemistry rules and their estimated kinetic parameters are available for a chemical system, RING can be used to identify a plausible set of dominant pathways that form specific products. One such case study is discussed in detail in <a name="bbib0295" href="#bib0295" class="workspace-trigger">Rangarajan et al. (in press)</a>, where we analyze the system of acid-catalyzed propane aromatization. In this example, we demonstrate three major features of RING – “encapsulation” of class of molecules using composite atom definitions, constraints-based extraction of pathways, and lumping of isomers leading to orders of magnitude reduction in network size.</p></dd><dt class="list-label">(b)</dt><dd class="list-description"><p id="par0200">Manual identification of all possible mechanisms exhaustively is cumbersome. Instead, RING can be used to generate and identify the mechanisms for the system. Specifically, if experimental data is available for a complex reaction network, RING can be used to test postulated elementary steps and propose a set of plausible pathways/mechanisms that describe how a product is formed in the system. In <a name="bbib0295" href="#bib0295" class="workspace-trigger">Rangarajan et al. (in press)</a>, an example involving the system of dehydration of glycerol to form acetone and a related system of acetone conversion to acetic acid is used for illustration. This example extensively uses the pathway identification feature of RING to query for specific pathways to products.</p></dd><dt class="list-label">(c)</dt><dd class="list-description"><p id="par0205">RING can be used to identify all possible synthetic routes to a compound from prescribed starting chemicals using specific chemistries. These routes can be pruned further by identifying those that satisfy properties like highest atom-efficiency and constraints like the length of the pathway or the presence/absence of certain types of reactions. In <a name="bbib0295" href="#bib0295" class="workspace-trigger">Rangarajan et al. (in press)</a>, we identify possible atom-efficient synthetic routes to form large alcohols from small oxygenates. Here, we use the complete mechanism identification feature of RING because the calculation of atom-efficiency requires the knowledge of the stoichiometry of the overall reaction.</p></dd></dl></p><p id="par0210">In summary, RING is a network generation and analysis tool developed using Cheminformatics and graph theoretical algorithms. It is composed of three components – a language compiler for a domain specific English-like reaction language, a reaction network generator that can construct an exhaustive network from specified reaction rules and initial reactants, and a post-processing module that enables (i) lumping of isomers to get a network of reduced size, (ii) finding pathways between initial reactants and specified products, and (iii) finding mechanisms for the formation of specified molecules in the network. RING can be used to construct reaction networks from elementary or overall reaction rules in a rule-based manner, and further, to elucidate the transformations occurring in complex chemical reaction networks through identifying reaction pathways, mechanisms, and lumps of isomers. The construction and analysis options in RING were described through an illustrative example involving acid catalyzed dehydration of fructose to HMF. Detailed case studies showcasing the network generation and analysis features of RING are discussed in the second part of this series (<a name="bbib0295" href="#bib0295" class="workspace-trigger">Rangarajan et al., in press</a>). RING is distributed open source under GNU Lesser GPL version 2.1 license through a dedicated webpage (<a name="bbib0315" href="#bib0315" class="workspace-trigger">RING, 2011</a>).</p></section></div><section id="aep-acknowledgment-id18"><h2 class="u-h3 u-margin-l-top u-margin-xs-bottom">Acknowledgments</h2><p id="par0220">The authors would like to thank Prof. Eric Van Wyk and Ted Kaminski, Department of Computer Science and Engineering, University of Minnesota, for helpful suggestions on computer science algorithms and collaboration on developing the reaction language for RING. Financial support from the Initiative for Renewable Energy (Large Grant: <a href="#gs0005">RL-0004-09</a>) at the <span id="gs0005">University of Minnesota</span>, the <span id="gs0010">National Science Foundation Emerging Frontiers in Research and Innovation program</span>, grant # <a href="#gs0010">0937706</a> is gratefully acknowledged. A.B. was supported as part of the Catalysis Center for Energy Innovation, an Energy Frontier Research Center funded by the <span id="gs0015">US Department of Energy, Office of Science, Office of Basic Energy Sciences</span> under Award number DE-SC0001004.</p></section><div class="Appendices"><section id="sec0160"><h2 class="u-h3 u-margin-l-top u-margin-xs-bottom">Appendix A. Supplementary data</h2><p id="par0230">The following are the Supplementary data to this article:<span class="display"><span class="e-component e-component-mmc1" id="upi0005"><span class="article-attachment"><a class="icon-link" href="https://ars.els-cdn.com/content/image/1-s2.0-S0098135412001846-mmc1.doc" title="Download Word document (112KB)" target="_blank" rel="noreferrer noopener"><svg focusable="false" viewBox="0 0 94 128" width="17.625" height="24" class="icon icon-text-document"><path d="m35.6 1e1c-5.38 0-10.62 1.92-14.76 5.4-9.1 7.68-18.84 20.14-18.84 32.1v70.5h9e1v-15.99-2.01-4e1 -17.64-32.36h-56.4zm0 1e1h46.4v22.36 17.64 4e1 2.01 5.99h-7e1v-49c0-6.08 4.92-11 11-11h17v-2e1h-6c-2.2 0-4 1.8-4 4v6h-7c-3.32 0-6.44 0.78-9.22 2.16 2.46-5.62 7.28-11.86 13.5-17.1 2.34-1.98 5.3-3.06 8.32-3.06zm-13.6 38v1e1h5e1v-1e1h-5e1zm0 2e1v1e1h5e1v-1e1h-5e1z"></path></svg></a><a class="anchor download-link u-font-sans" href="https://ars.els-cdn.com/content/image/1-s2.0-S0098135412001846-mmc1.doc" target="_blank" download="" title="Download Word document (112KB)"><span class="anchor-text">Download : <span class="download-link-title">Download Word document (112KB)</span></span></a></span></span></span></p></section></div></div>